### PR TITLE
Add Fee for limit orders

### PR DESCRIPTION
**Pull request by anxolin**
# Summary

<<if there's an issue>>Fixes #issueNumber

*High-level description of what your changes are accomplishing*

*Add screenshots if applicable. Images are nice :)*

  # To Test

1. <<Step one>> Open the page `about`
- [ ] <<What to expect?>> Verify it contains about information...
- [ ] Checkbox Style list of things a QA person could verify, i.e.
- [ ] Should display Text Input our storybook
- [ ] Input should not accept Numbers
2. <<Step two>> ...

  # Background

*Optional: Give background information for changes you've made, that might be difficult to explain via comments*

### CODE REVIEW

File: src/cow-react/abis/types/index.ts [hunk 1]
```diff
@@ -10,5 +10,5 @@ export { GPv2Settlement__factory } from "./factories/GPv2Settlement__factory";
 export { MerkleDrop__factory } from "./factories/MerkleDrop__factory";
 export { TokenDistro__factory } from "./factories/TokenDistro__factory";
 export { VCow__factory } from "./factories/VCow__factory";
-export * from "@src/abis/types";
 export * from "@cow/abis/types/ethflow";
+export * from "@src/abis/types";
```
Comment (nlordell) [inline src/cow-react/abis/types/index.ts:14]: Got to keep those imports alphabetically ordered 😛.

File: src/cow-react/modules/limitOrders/pure/ReceiptModal/FeeField.tsx [hunk 2]
```diff
@@ -6,15 +6,17 @@ import * as styledEl from './styled'
...
 
 export function FeeField({ order }: Props): JSX.Element | null {
-  const { executedFeeAmount, inputToken, sellToken } = order
+  const { executedFeeAmount, executedSurplusFee, inputToken, sellToken } = order
 
-  let formattedExecutedFee: string | undefined = executedFeeAmount
+  let formattedExecutedFee: string | undefined = executedSurplusFee || executedFeeAmount
   let quoteSymbol: string | undefined = sellToken
-  let executedFeeCurrency: CurrencyAmount<Token> | undefined
+  let totalFeel: CurrencyAmount<Token> | undefined
```
Comment (nlordell) [inline src/cow-react/modules/limitOrders/pure/ReceiptModal/FeeField.tsx:13]: spelling nit:
```suggestion
  let totalFee: CurrencyAmount<Token> | undefined
```

File: src/cow-react/modules/limitOrders/pure/ReceiptModal/FeeField.tsx [hunk 3]
```diff
@@ -6,15 +6,17 @@ import * as styledEl from './styled'
...
 
   if (sellToken) {
-    executedFeeCurrency = CurrencyAmount.fromRawAmount(inputToken, executedFeeAmount || 0)
-    formattedExecutedFee = formatSmart(executedFeeCurrency)
+    const executedFeeCurrency = CurrencyAmount.fromRawAmount(inputToken, executedFeeAmount || 0)
+    const executedSurplusFeeCurrency = CurrencyAmount.fromRawAmount(inputToken, executedSurplusFee || 0)
+    totalFeel = executedFeeCurrency.add(executedSurplusFeeCurrency)
```
Comment (nlordell) [inline src/cow-react/modules/limitOrders/pure/ReceiptModal/FeeField.tsx:18]: This is a bit inconsistent with ☝️ (which was "use `executedSurplusFee` if its there, and `executedFeeAmount` otherwise).

I actually think that it should just be:
```ts
totalFee = CurrencyAmount.fromRawAmount(inputToken, formattedExecutedFee)
```

The reason is that the `executedSurplusFee` represents exactly the fee that was charged (regardless of the fee signed with the order).

So, while the protocol currently does not allow placing a `limit` order with any other fee than `0` - the backend is designed to support these kinds of orders for the future. For instance, imagine a limit order with:
```
sellAmount: 100
buyAmount:   90
feeAmount:   10
surplusFee:   5
```

If I were to manually insert this order into the orderbook database, the "synthetic" order that gets created and used for solving is:
```
sellAmount: 105 // = sellAmount + feeAmount - surplusFee
buyAmount:   90
feeAmount:    5 // = surplusFee
```

And so, the solvers would use some of the sell token from the `feeAmount` for actual trading (even if you would over-estimate the `feeAmount` value).

While, as I previously mentioned, we don't allow this ATM - this is the logic that the backend uses for limit orders.

**tl;dr:** the "real" fee is `executedSurplusFee ?? executedFeeAmount`, and _not_ `(executedSurplusFee ?? 0) + executedFeeAmount`.

**Review by elena-zh — APPROVED** [review]
LGTM!
Fees are displayed, amounts match the ones on Explorer dev. 
But don't match amounts that are in https://github.com/cowprotocol/explorer/pull/265#issuecomment-1335112655 PR.

File: src/cow-react/api/gnosisProtocol/api.ts [hunk 4]
```diff
@@ -105,6 +105,7 @@ export interface OrderMetaData {
   executedSellAmount: string
   executedSellAmountBeforeFees: string
   executedFeeAmount: string
+  executedSurplusFee: string
```
Comment (shoom3301) [inline src/cow-react/api/gnosisProtocol/api.ts:108]: `executedSurplusFee: string | null`

Sorry, maybe I forgot, but, why don't we use the interface from the SDK?

Comment (shoom3301) [reply src/cow-react/api/gnosisProtocol/api.ts:108]: P.s. added `totalFee` to SDK:
https://github.com/cowprotocol/cow-sdk/pull/88

File: src/cow-react/modules/limitOrders/pure/ReceiptModal/FeeField.tsx (hunk 2 above)
Comment (anxolin) [reply src/cow-react/modules/limitOrders/pure/ReceiptModal/FeeField.tsx:13]: 🙈 
image

File: src/cow-react/modules/limitOrders/pure/ReceiptModal/FeeField.tsx (hunk 3 above)
Comment (anxolin) [reply src/cow-react/modules/limitOrders/pure/ReceiptModal/FeeField.tsx]: Thanks for reviewing this and explaining with such details! It helps to understand your point. 

I'm not sure I understand, because the surplus fee is actually unknown until only after the settlement is executed (after all, the `executed` is part of the fee name). 

This means that the only difference between your proposal and my proposal will be after execution only. 

Then, if I get the example correctly, a user in the future could include a fee, so there's less fee to take from the surplus. For example, a website (or us if we add the setting) could include as fee the current fee, so in the case of the fee being roughly the same as in placing-order time, then the surplusFee will be roughly 0. Nice feature --> if you want to trade closer to the target price. 

So, in the example, if the user signed 10 as the fee, and we need:
- `8` --> we take 10
- `10` --> nice! we take 10
- `12` ---> we take 10 + 2 of surplus fee

For this example, I would show `8`, `10` and `12` respectively (sum of both), instead of `10`, `10` and `2` that I understand would be if we  use `executedSurplusFee ?? executedFeeAmount` 

Does this make sense?

File: src/cow-react/api/gnosisProtocol/api.ts (hunk 4 above)
Comment (anxolin) [reply src/cow-react/api/gnosisProtocol/api.ts:108]: Nice! 

> executedSurplusFee: string | null

Why nullable?  As I see in the spec, it looks like is mandatory
https://api.cow.fi/docs/#/default/get_api_v1_orders__UID_

image

Comment (shoom3301) [reply src/cow-react/api/gnosisProtocol/api.ts:108]: By fact it's nullable :) CHeck the real response from API for market orders

File: src/cow-react/modules/limitOrders/pure/ReceiptModal/FeeField.tsx (hunk 3 above)
Comment (nlordell) [reply src/cow-react/modules/limitOrders/pure/ReceiptModal/FeeField.tsx]: > So, in the example, if the user signed 10 as the fee, and we need:
>
> - `8` --> we take 8, we add 2 to the sell amount and trade a little more with an AMM in order to give the user some extra surplus
> - `10` --> nice! we take 10
> - `12` ---> we take 10 + 2 of surplus fee

File: src/cow-react/api/gnosisProtocol/api.ts (hunk 4 above)
Comment (nlordell) [reply src/cow-react/api/gnosisProtocol/api.ts:108]: Oops! Looks like an OpenAPI spec mistake. It is indeed nullable.

File: src/cow-react/modules/limitOrders/pure/ReceiptModal/FeeField.tsx (hunk 3 above)
Comment (nlordell) [reply src/cow-react/modules/limitOrders/pure/ReceiptModal/FeeField.tsx]: > then the surplusFee will be roughly 0

`surplusFee` is always the total fee that needs to be taken from the limit order irrespective of the `feeAmount` that is signed with the order.

Comment (anxolin) [reply src/cow-react/modules/limitOrders/pure/ReceiptModal/FeeField.tsx]: > surplusFee is always the total fee that needs to be taken from the limit order irrespective of the feeAmount that is signed with the order.

Ok, I see now! it just feels counter-intuitive to me. Imo it should be called `totalExecutedFee` or it should be only the part that needed to be taken from the surplus on top of the fee the user agreed to pay.

File: src/cow-react/api/gnosisProtocol/api.ts (hunk 4 above)
Comment (anxolin) [reply src/cow-react/api/gnosisProtocol/api.ts:108]: > By fact it's nullable :) CHeck the real response from API for market orders

refuted by evidence :P

**Review by shoom3301 — APPROVED** [review]

**Comment by elena-zh** [comment]
Changes LGTM!
